src/Particle/Lattice/CrystalLattice.cpp

```diff
@@ -38,6 +38,25 @@ CrystalLattice<T, D>::CrystalLattice()
   reset();
 }
 
+template<typename T, unsigned D>
+template<typename TT>
+CrystalLattice<T, D>::CrystalLattice(const Tensor<TT, D>& lat)
```

Comment:
Is it beneficial to have this constructor? Right now this is a short cut of `CrystalLattice()` + `set`. Due to the way how we parse xml and setup `CrystalLattice` I feel `CrystalLattice()` + `set` is simple enough and sufficient and prefer not to add another code path.

Reply:
I also don't like the two pass construction. I also prefer to have a single constructor taking care of all the cases and in this way, I can make a lot of things const safely.

> currently some test code omits calling reset() after manipulating internals which means many elements of the lattice are invalid and hopefully not relevant to the test while providing potentially fatal example code.

I hate those examples. Internals should not be modified and input argument should be minimal and validated.

I just feel it is not a good time to tackle them. When we really tackle them, we should just make input arguments in full precision only and keep constructors simple.